AFQ/tasks/data.py

```diff
@@ -285,22 +288,131 @@ def csd(dwi, brain_mask, gtab, data,
     return csdf.shm_coeff, meta
 
 
-csd_params = pimms.calc("csd_params")(csd)
-
-
-@pimms.calc("pmap")
+@pimms.calc("csd_pmap")
 @as_file(suffix='_model-CSD_desc-APM_dwi.nii.gz')
+@as_img
 def anisotropic_power_map(csd_params):
     """
     full path to a nifti file containing
     the anisotropic power map
     """
-    sh_coeff = nib.load(csd_params)
-    pmap = shm.anisotropic_power(sh_coeff.get_fdata())
-    pmap = nib.Nifti1Image(pmap, sh_coeff.affine)
+    sh_coeff = nib.load(csd_params).get_fdata()
+    pmap = shm.anisotropic_power(sh_coeff)
     return pmap, dict(CSDParamsFile=csd_params)
 
 
+@pimms.calc("csd_ai")
+@as_file(suffix='_model-CSD_desc-AI_dwi.nii.gz')
+@as_img
+def anisotropic_index(csd_params):
+    """
+    full path to a nifti file containing
+    the anisotropic index
+    """
+    sh_coeff = nib.load(csd_params).get_fdata()
+    sh_0 = sh_coeff[..., 0] ** 2
+    sh_sum_squared = np.sum(sh_coeff ** 2, axis=-1)
+    AI = np.zeros_like(sh_0)
+    AI = np.sqrt(1 - sh_0 / sh_sum_squared)
+    return AI, dict(CSDParamsFile=csd_params)
+
+
+@pimms.calc("gq_params", "gq_iso", "gq_aso")
+def gq(base_fname, gtab, dwi_affine, data,
+       gq_sampling_length=1.2):
+    """
+    full path to a nifti file containing
+    parameters for the Generalized Q-Sampling
+    shm_coeff,
+    full path to a nifti file containing isotropic diffusion component,
+    full path to a nifti file containing anisotropic diffusion component
+    Parameters
+    ----------
+    gq_sampling_length : float
+        Diffusion sampling length.
+        Default: 1.2
+    """
+    gqmodel = GeneralizedQSamplingModel(
+        gtab,
+        sampling_length=gq_sampling_length)
+
+    gqi_vector = np.real(
+        squared_radial_component(np.dot(
+            gqmodel.b_vector, default_sphere.vertices.T)
+            * gqmodel.Lambda))
+    ODF = blas.dgemm(
```

Comment:
add a test here that it produces the same result as in dipy.

Reply:
Once you figure out why the dipy version takes too long, start a PR to change this in DIPY